### ISSUE TITLE
aio_write() blocks reactor thread for 3.3 sec on xfs_ilock with kworkers

### ISSUE THREAD

**Issue by tgrabiec** [scylladb/seastar#341]
Probably related to concurrent fdatasync (`file::flush()`).

Refs https://github.com/scylladb/scylla/issues/2127

[Setup details here](https://gist.github.com/tgrabiec/0ef5779e85d3551047ac6c67515dc013/bf108a13075dc41a5efb25ab78572a88b029e7f9)

```
perf_fast_forwa  1632 [000]   396.404031: sched:sched_switch: perf_fast_forwa:1632 [120] D ==> timer-0:1634 [120]
            7fff9287472f __schedule ([kernel.kallsyms])
            7fff92874d16 schedule ([kernel.kallsyms])
            7fff92878491 rwsem_down_write_failed ([kernel.kallsyms])
            7fff9242cd27 call_rwsem_down_write_failed ([kernel.kallsyms])
            7fff92877a3d down_write ([kernel.kallsyms])
            7fffc05d2fe3 xfs_ilock ([kernel.kallsyms])
            7fffc05d01db xfs_vn_update_time ([kernel.kallsyms])
            7fff92285ade file_update_time ([kernel.kallsyms])
            7fffc05c3dff xfs_file_aio_write_checks ([kernel.kallsyms])
            7fffc05c405c xfs_file_dio_aio_write ([kernel.kallsyms])
            7fffc05c456d xfs_file_write_iter ([kernel.kallsyms])
            7fff922bc5d3 aio_write ([kernel.kallsyms])
            7fff922bcec1 do_io_submit ([kernel.kallsyms])
            7fff922bdd40 sys_io_submit ([kernel.kallsyms])
            7fff9287a6b7 entry_SYSCALL_64_fastpath ([kernel.kallsyms])
                     687 io_submit (/usr/lib64/libaio.so.1.0.1)
                  112373 seastar::reactor::flush_pending_aio (/home/tgrabiec/src/scylla/build/release/tests/perf/perf_fast_forward_g)
```

Woken up:
```
kworker/0:1   137 [000]   399.725111: sched:sched_wakeup: perf_fast_forwa:1632 [120] success=1 CPU:000
            7fff920d11b9 ttwu_do_wakeup ([kernel.kallsyms])
            7fff920d12d7 ttwu_do_activate ([kernel.kallsyms])
            7fff920d1f00 try_to_wake_up ([kernel.kallsyms])
            7fff920d21f4 wake_up_q ([kernel.kallsyms])
            7fff920f624e rwsem_wake ([kernel.kallsyms])
            7fff9242cd8b call_rwsem_wake ([kernel.kallsyms])
            7fff920f4a15 up_write ([kernel.kallsyms])
            7fffc05d3314 xfs_iunlock ([kernel.kallsyms])
            7fffc05cfc6d xfs_iomap_write_unwritten ([kernel.kallsyms])
            7fffc05c31ef xfs_dio_write_end_io ([kernel.kallsyms])
            7fff922d3cd0 iomap_dio_complete ([kernel.kallsyms])
            7fff922d3d65 iomap_dio_complete_work ([kernel.kallsyms])
            7fff920c0593 process_one_work ([kernel.kallsyms])
            7fff920c080a worker_thread ([kernel.kallsyms])
            7fff920c6e55 kthread ([kernel.kallsyms])
            7fff9287a905 ret_from_fork ([kernel.kallsyms])
```

There are multiple kworkers active, blocking each other on xfs_ilock and buf lock. Example kworker wake point:
```
kworker/0:1H   631 [000]   399.724993: sched:sched_wakeup: kworker/0:0:3 [120] success=1 CPU:000
            7fff920d11b9 ttwu_do_wakeup ([kernel.kallsyms])
            7fff920d12d7 ttwu_do_activate ([kernel.kallsyms])
            7fff920d1f00 try_to_wake_up ([kernel.kallsyms])
            7fff920d21a5 wake_up_process ([kernel.kallsyms])
            7fff92877678 __up.isra.0 ([kernel.kallsyms])
            7fff920f4856 up ([kernel.kallsyms])
            7fffc05bea0a xfs_buf_unlock ([kernel.kallsyms])
            7fffc05ea06c xfs_buf_item_unpin ([kernel.kallsyms])
            7fffc05e2313 xfs_trans_committed_bulk ([kernel.kallsyms])
            7fffc05e7527 xlog_cil_committed ([kernel.kallsyms])
            7fffc05e4199 xlog_state_do_callback ([kernel.kallsyms])
            7fffc05e435d xlog_state_done_syncing ([kernel.kallsyms])
            7fffc05e4400 xlog_iodone ([kernel.kallsyms])
            7fffc05beadb xfs_buf_ioend ([kernel.kallsyms])
            7fffc05bec25 xfs_buf_ioend_work ([kernel.kallsyms])
            7fff920c0593 process_one_work ([kernel.kallsyms])
            7fff920c080a worker_thread ([kernel.kallsyms])
            7fff920c6e55 kthread ([kernel.kallsyms])
            7fff9287a905 ret_from_fork ([kernel.kallsyms])

```

**Comment by avikivity** [comment]
Linux 4.13 supports nowait AIO, that's supposed to return EWOULDBLOCK in such cases (and we have to punt to the syscall thread).

Does `--unsafe-bypass-fsync` help?

**Comment by nyh** [comment]
@tgrabiec how did you conclude the problem is fdatasync()? Looking at the first backtrace you showed, I see file_update_time(), which is specifically not done by fdatasync()...

I'm also curious what is the "rwsem_down_write_failed()" in the first backtrace.

**Comment by avikivity** [comment]
`fdatasync` likely locks the inode, so writes block on it.

**Comment by avikivity** [comment]
failed = did not acquire the lock, so must yield to other processes

**Comment by tgrabiec** [comment]
@nyh file_update_time() is done by aio_write, but it blocks on `xfs_ilock`, which is held by kworkers, which I assumed are doing some work on behalf of `fdatasync`. I saw in other traces that `fdatasync()` does in-core log flushing via kworkers.

**Comment by avikivity** [comment]
I wonder why we have a concurrent write + fdatasync. I don't think we ever call fdatasync prematurely.

**Comment by tgrabiec** [comment]
@avikivity I still see sleeps in aio_write with syncs disabled, though not as long (up to 1s).


```
perf_fast_forwa  3668 [000]  9697.257642: sched:sched_switch: perf_fast_forwa:3668 [120] D ==> swapper/0:0 [120]
            7fff9287472f __schedule ([kernel.kallsyms])
            7fff92874d16 schedule ([kernel.kallsyms])
            7fff92878491 rwsem_down_write_failed ([kernel.kallsyms])
            7fff9242cd27 call_rwsem_down_write_failed ([kernel.kallsyms])
            7fff92877a3d down_write ([kernel.kallsyms])
            7fffc05d2fe3 xfs_ilock ([kernel.kallsyms])
            7fffc05d01db xfs_vn_update_time ([kernel.kallsyms])
            7fff92285ade file_update_time ([kernel.kallsyms])
            7fffc05c3dff xfs_file_aio_write_checks ([kernel.kallsyms])
            7fffc05c405c xfs_file_dio_aio_write ([kernel.kallsyms])
            7fffc05c456d xfs_file_write_iter ([kernel.kallsyms])
            7fff922bc5d3 aio_write ([kernel.kallsyms])
            7fff922bcec1 do_io_submit ([kernel.kallsyms])
            7fff922bdd40 sys_io_submit ([kernel.kallsyms])
            7fff9287a6b7 entry_SYSCALL_64_fastpath ([kernel.kallsyms])
                     687 io_submit (/usr/lib64/libaio.so.1.0.1)

```

```
kworker/0:1H   631 [000]  9698.059717: sched:sched_wakeup: kworker/0:3:236 [120] success=1 CPU:000
            7fff920d11b9 ttwu_do_wakeup ([kernel.kallsyms])
            7fff920d12d7 ttwu_do_activate ([kernel.kallsyms])
            7fff920d1f00 try_to_wake_up ([kernel.kallsyms])
            7fff920d21a5 wake_up_process ([kernel.kallsyms])
            7fff92877678 __up.isra.0 ([kernel.kallsyms])
            7fff920f4856 up ([kernel.kallsyms])
            7fffc05bea0a xfs_buf_unlock ([kernel.kallsyms])
            7fffc05ea06c xfs_buf_item_unpin ([kernel.kallsyms])
            7fffc05e2313 xfs_trans_committed_bulk ([kernel.kallsyms])
            7fffc05e7527 xlog_cil_committed ([kernel.kallsyms])
            7fffc05e4199 xlog_state_do_callback ([kernel.kallsyms])
            7fffc05e435d xlog_state_done_syncing ([kernel.kallsyms])
            7fffc05e4400 xlog_iodone ([kernel.kallsyms])
            7fffc05beadb xfs_buf_ioend ([kernel.kallsyms])
            7fffc05bec25 xfs_buf_ioend_work ([kernel.kallsyms])
            7fff920c0593 process_one_work ([kernel.kallsyms])
            7fff920c080a worker_thread ([kernel.kallsyms])
            7fff920c6e55 kthread ([kernel.kallsyms])
            7fff9287a905 ret_from_fork ([kernel.kallsyms])

:3680  3680 [000]  9698.187978: sched:sched_wakeup: perf_fast_forwa:3668 [120] success=1 CPU:000
            7fff920d11b9 ttwu_do_wakeup ([kernel.kallsyms])
            7fff920d12d7 ttwu_do_activate ([kernel.kallsyms])
            7fff920d1f00 try_to_wake_up ([kernel.kallsyms])
            7fff920d21f4 wake_up_q ([kernel.kallsyms])
            7fff920f624e rwsem_wake ([kernel.kallsyms])
            7fff9242cd8b call_rwsem_wake ([kernel.kallsyms])
            7fff920f4a15 up_write ([kernel.kallsyms])
            7fffc05d3314 xfs_iunlock ([kernel.kallsyms])
            7fffc05cfc6d xfs_iomap_write_unwritten ([kernel.kallsyms])
            7fffc05c31ef xfs_dio_write_end_io ([kernel.kallsyms])
            7fff922d3cd0 iomap_dio_complete ([kernel.kallsyms])
            7fff922d3d65 iomap_dio_complete_work ([kernel.kallsyms])
            7fff920c0593 process_one_work ([kernel.kallsyms])
            7fff920c080a worker_thread ([kernel.kallsyms])
            7fff920c6e55 kthread ([kernel.kallsyms])
            7fff9287a905 ret_from_fork ([kernel.kallsyms])

```

syscall stats from another run:


```
 syscall-0 (3807), 11070 events, 0.0%

   syscall            calls    total       min       avg       max      stddev
                               (msec)    (msec)    (msec)    (msec)        (%)
   --------------- -------- --------- --------- --------- ---------     ------
   read                1302 104714.781     0.000    80.426 21907.504     31.75%
   ftruncate            306   618.570     0.034     2.021    15.142     10.18%
   open                 508   128.138     0.002     0.252    13.046     26.28%
   unlink               198     3.234     0.004     0.016     0.043      3.57%
   tgkill               695     1.516     0.001     0.002     0.013      1.99%
   stat                 396     1.233     0.002     0.003     0.014      2.39%
   ioctl                396     1.033     0.001     0.003     0.015      2.03%
   close                550     0.872     0.001     0.002     0.005      1.64%
   getpid               695     0.772     0.001     0.001     0.003      1.24%
   rename                56     0.725     0.006     0.013     0.036      5.09%
   fcntl                396     0.593     0.001     0.001     0.004      1.36%
   mkdir                  2     0.125     0.057     0.062     0.068      9.25%
   fstat                 34     0.054     0.001     0.002     0.003      3.68%
   madvise                1     0.046     0.046     0.046     0.046      0.00%


 timer-0 (3806), 20362 events, 0.0%

   syscall            calls    total       min       avg       max      stddev
                               (msec)    (msec)    (msec)    (msec)        (%)
   --------------- -------- --------- --------- --------- ---------     ------
   read               10178 105821.606     0.000    10.397 21203.804     28.14%
   mmap                   1     0.012     0.012     0.012     0.012      0.00%
   madvise                2     0.006     0.002     0.003     0.004     39.13%


 perf_fast_forwa (3805), 62073177 events, 99.9%

   syscall            calls    total       min       avg       max      stddev
                               (msec)    (msec)    (msec)    (msec)        (%)
   --------------- -------- --------- --------- --------- ---------     ------
   timerfd_settime 16314188 20574.018     0.001     0.001     0.019      0.01%
   epoll_pwait     14530172 17834.772     0.001     0.001     1.101      0.02%
   io_submit          81829  5575.203     0.002     0.068   147.462     11.23%
   io_getevents       87222   157.249     0.001     0.002     0.033      0.15%
   timer_settime      10432    22.299     0.001     0.002     0.035      0.40%
   write               2055     4.592     0.001     0.002     0.013      1.13%
   rt_sigprocmask      1950     2.886     0.001     0.001     0.018      1.31%
   ioctl               1016     1.389     0.001     0.001     0.064      4.70%
   ftruncate            268     1.292     0.002     0.005     0.022      2.73%
   lstat                553     1.034     0.001     0.002     0.013      1.63%
   fcntl                508     0.549     0.001     0.001     0.012      2.32%
   lseek                362     0.483     0.001     0.001     0.013      3.56%
   fstat                309     0.437     0.001     0.001     0.013      3.16%
   futex                  3     0.414     0.003     0.138     0.291     60.50%
   stat                  66     0.175     0.002     0.003     0.012      7.65%
   getcwd                66     0.120     0.001     0.002     0.003      2.64%
   close                 31     0.070     0.001     0.002     0.005      9.73%
   unlink                 2     0.014     0.006     0.007     0.008     16.20%
   munmap                 1     0.010     0.010     0.010     0.010      0.00%
   rmdir                  3     0.009     0.002     0.003     0.004     17.23%
   getdents               6     0.009     0.001     0.001     0.004     27.42%
   open                   3     0.008     0.002     0.003     0.004     25.75%
   epoll_ctl              1     0.005     0.005     0.005     0.005      0.00%
   sigaltstack            1     0.003     0.003     0.003     0.003      0.00%
   timer_delete           1     0.002     0.002     0.002     0.002      0.00%


```

**Comment by nyh** [comment]
@avikivity don't we call call fdatasync() (which we call, sadly, flush()), at the end while the last writes are still in progress (the iosubmit returned but not the actual operation)?

**Comment by avikivity** [comment]
@nyh no (it would be meaningless)

**Comment by avikivity** [comment]
@tgrabiec the second trace is not for the inode lock, but an internal xfs buffer. The third is problematic. Does the `relatime` mount option help?

**Comment by avikivity** [comment]
Scratch that, `relatime` would not help.

**Comment by nyh** [comment]
@avikivity I mean something like the following in core/iostream-impl.hh:
```c++
            return put(std::move(_buf)).then([this] {
                return _fd.flush();
```
Doeesn't this iosubmit() a write and then immediately does fdatasync? Maybe I'm misunderstanding something.

**Comment by nyh** [comment]
By the way, it seems XFS internally has a file mode FMODE_NOCMTIME which together with open's MODE_NOATIME makes sure it never bothers with that inode-locking file timestamp crap. But I don't know how to enable FMODE_NOCMTIME. I saw on the Linux kernel mailing list a bunch of patches to add "O_NOCMTIME" and "O_NOMTIME" but I don't know what ever came out of it.

**Comment by avikivity** [comment]
It looks like the xfs log is serializing on its internal fsync.

```
write 1 completes
 -> update extent map
    -> lock inode
        -> wait for previous transaction fsync to complete
write 2 completes
 -> update mtime
    -> lock inode
        -> blocks
```

**Comment by nyh** [comment]
@tgrabiec so it's probably not related to fdatasync() at all... If we just have a lot of writes, if something (probably what Avi just found) locks an inode for a whole second, then any io_submit() of a write will block while it locks the inode for the idiotic reason of updating the time.

Disabling the mtime modification (a good question is how - see what I wrote above) could avoid this lock. I don't know if there's another way (can we avoid the inode lock during the long one-second operation?).

**Comment by tgrabiec** [comment]
@nyh The problem also happens without fdatasync, so it's not the only source of blocking, but fdatasync is such a source, because with it enabled the pauses are a lot larger.

**Comment by nyh** [comment]
Proposed patch to add O_NOCMTIME open option in Linux:
https://groups.google.com/d/msg/linux.kernel/ajyGLikokMU/J-zCyu5aJfwJ
Sadly it seems it was never committed, nor even discussed.

**Comment by avikivity** [comment]
It will probably happen without mtime updates too. Every write completion updates the extent map (which can take the inode lock and block if the xfs log has a large backlog), and write initiations take the inode lock as well.

**Comment by avikivity** [comment]
Perhaps this will help:

```
diff --git a/fs/direct-io.c b/fs/direct-io.c
index 5fa2211e49ae..91bc063ee85b 100644
--- a/fs/direct-io.c
+++ b/fs/direct-io.c
@@ -1197,11 +1197,11 @@ do_blockdev_direct_IO(struct kiocb *iocb, struct inode *inode,
                dio->is_async = true;
 
        dio->inode = inode;
        if (iov_iter_rw(iter) == WRITE) {
                dio->op = REQ_OP_WRITE;
-               dio->op_flags = REQ_SYNC | REQ_IDLE;
+               dio->op_flags = REQ_IDLE;
                if (iocb->ki_flags & IOCB_NOWAIT)
                        dio->op_flags |= REQ_NOWAIT;
        } else {
                dio->op = REQ_OP_READ;
        }
```

The theory is that REQ_SYNC writes prevent xfs metadata writes from bypassing normal data writes, so they block the whole filesystem.

**Comment by avikivity** [comment]
btw, reducing `--max-io-requests` may help. We measure it for 4k reads and then apply it for 128k writes.

**Comment by avikivity** [comment]
Fixed by 0fb334b9f1f0d624e256436d2c1e4839d7eff2dd. Fix will improve as XFS is patched to allow concurrent fsync/write.
